### DISCUSSION TITLE
How best to test the __resolveReference resolver?

### DISCUSSION THREAD

**Discussion by tjmgregory** [apollographql/apollo-server#5866] [Q&A] [answered]
Howdy folks!

As part of our federated gateway, we use the `__resolveReference` approach to allow service `api-books` (I'm using fake names/types for ease) to have an entity `Book` which has a field `author: Author`, where `Author` is an entity from service `api-authors`.

To accomplish this, in `api-authors`, I have the following:
```typescript
import { Author, Resolvers, GraphQLRecursivePick } from '../graphql/types'

const resolvers: Resolvers = {
  Author: {
    async __resolveReference(
      { id }: GraphQLRecursivePick<Author, { id: true }>,
    ): Promise<Author> {
      ...
    }
}
 ```

How should one test (ideally with the service running, so not unit) this resolver?

Now, no other entity in `api-authors` has an `author: Author` field, so we can't hop on the back of an integration test that returns that entity, and include `author` in the return fields that way.

Ideally too I would prefer not having to add types/resolvers to accomplish this for the sole purpose of testing, as this gives me a less confidence that this would work in production.

What do folks suggest? :) Thanks all!

Edit: Should add that I've scoured high and low for suggestions in the docs and online, but have come up dry 💧

**Comment by lennyburdette** [comment, marked as answer]
You can pretend to be the gateway and executing the `__resolveReference` functions by calling the `_entities` field like this:

```graphql
query ($representations: [_Any!]!) {
  _entities(representations: $representations) {
    ... on User {
      id
      username
    }
  }
}
```

which is documented here: https://www.apollographql.com/docs/federation/subgraphs/#query_entities

Hope this helps!

**Reply by tjmgregory** [reply]
Oh sensational, this is spot on what I have been looking for! Thank you so much, you're a hero 🦸 💪 I'd even scanned over this section but somehow didn't connect the dots 🤦 
Thanks again!

**Reply by brackendawson** [reply]
The documentation linked above seems to have changed, the example is now here: https://www.apollographql.com/docs/graphos/platform/production-readiness/testing-with-apollo-federation#entity-resolvers-in-integration-tests
